lib/package.js

```diff
@@ -760,24 +761,40 @@ exports.createMain = function(pkg, pjson, downloadDir) {
     }
 
     if (main) {
-      if (main.startsWith('./'))
-        main = main.substr(2);
-      if (main.endsWith('.js'))
-        main = main.substr(0, main.length - 3);
+      if (main.startsWith('./')) {
+        main = main.substr(2);    
+      }
+    }
+    else {
+      main = 'index' + mainFileExtension;
     }
 
-    main = main || 'index';
+    return new Promise(function(resolve) {
+      // Check to see if main points to an actual file or not
+      // i.e. does it include extension.
+      mainPath = path.resolve(downloadDir, main);
+      fs.exists(mainPath, resolve);
+    });
+
+  }).then(function(exists) {
+    if (exists) {
+      var match = /(\.[^.]+)$/.exec(main);
+      if (match) {
+        mainFileExtension = match[1];
+        main = main.slice(0, - mainFileExtension.length);
```

Comment:
That's what the original code does - removing `.js` from `main`. But as I read through the code, I agree with you. Keeping the extension will be a better solution, as the user can do:

````
packages: {
  "myModule": {
    defaultExtension: false
  }
}
````

which will be more aligned with 0.17, correct?

I'll make the change accordingly.